utils/class-capability-utils.php

```diff
@@ -34,15 +43,70 @@ public static function user_has_elevated_permissions( $user, $capabilities = [],
 		$roles        = self::normalize_roles_input( $roles );
 		
 		if ( ! empty( $capabilities ) ) {
-			return self::user_has_any_capability( $user, $capabilities );
+			// Always use full resolution for capabilities (detects dynamic grants)
+			return self::user_has_any_capability_full( $user, $capabilities );
 		}
 		
 		return self::user_has_any_role( $user, $roles );
 	}
 	
+	/**
+	 * Check if a user has any of the specified capabilities using full capability resolution.
+	 * 
+	 * This method uses user_can() to check capabilities, which includes dynamic grants
+	 * from map_meta_cap and user_has_cap filters. It includes recursion protection
+	 * to prevent infinite loops when called from within capability filters.
+	 * 
+	 * @param \WP_User|false|null $user The user to check.
+	 * @param array $capabilities Array of capabilities to check.
+	 * @return bool True if user has any of the capabilities, false otherwise.
+	 */
+	public static function user_has_any_capability_full( $user, $capabilities ): bool {
+		if ( ! ( $user instanceof \WP_User ) || ! $user->exists() || empty( $capabilities ) ) {
+			return false;
+		}
+		
+		// Check if we're already checking capabilities for this user
+		if ( isset( self::$checking_capability[ $user->ID ] ) && self::$checking_capability[ $user->ID ] ) {
+			// We're in a recursive call - fall back to allcaps check
+			Logger::error(
+				self::LOG_FEATURE_NAME,
+				'Recursion detected in user_has_any_capability_full for user ID ' . $user->ID
+			);
+				return self::user_has_any_capability( $user, $capabilities );
+		}
+		
+		// Set recursion guard
+		self::$checking_capability[ $user->ID ] = true;
+		
+		$has_capability = false;
+		
+		try {
+			foreach ( $capabilities as $capability ) {
+				if ( ! is_scalar( $capability ) ) {
```

Comment:
I think this will still allow empty strings to pass.
Stricter check:
```suggestion
				if ( ! is_string( $capability ) || trim( $capability ) === '' ) {
```

Reply:
Fixed in [5bee824](https://github.com/Automattic/vip-security-boost/pull/76/commits/5bee8247fe6f0aed6f54a5bebeb4fb12aaaf6f19)!